src/libsyntax/tokenstream.rs

```diff
@@ -146,13 +144,14 @@ impl TokenTree {
 #[derive(Clone, Debug)]
 pub enum TokenStream {
     Empty,
-    Tree(TokenTree, IsJoint),
-    Stream(Lrc<Vec<TokenStream>>),
+    Stream(Lrc<Vec<TreeAndJoint>>),
 }
```

Comment:
It would probably be clearer to use what `ThinTokenStream` used previously instead of `enum`.
```rust
pub struct TokenStream(Option<Lrc<Vec<TreeAndJoint>>>);
```
(And also move comment explaining the `Option` from the removed `ThinTokenStream` to here.)

Reply:
This will make `Option` helpers like `map` etc usable as well.

Reply:
LGTM beside this one comment.